onnxruntime/test/testdata/custom_op_openvino_wrapper_library/custom_op_lib.cc

```diff
@@ -0,0 +1,45 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <mutex>
+#include <vector>
+
+#include "custom_op_lib.h"
+#include "openvino_wrapper.h"
+#include "core/common/common.h"
+
+static const char* c_OpDomain = "test.customop.ov";
+
+static void AddOrtCustomOpDomainToContainer(Ort::CustomOpDomain&& domain) {
+  static std::vector<Ort::CustomOpDomain> ort_custom_op_domain_container;
+  static std::mutex ort_custom_op_domain_mutex;
+  std::lock_guard<std::mutex> lock(ort_custom_op_domain_mutex);
+  ort_custom_op_domain_container.push_back(std::move(domain));
+}
+
+OrtStatus* ORT_API_CALL RegisterCustomOps(OrtSessionOptions* options, const OrtApiBase* api_base) {
+
+  // Allow use of Ort::GetApi() and Ort::OpWrapper::GetApi() in C++ ORT api implementations.
+  Ort::InitApi(api_base->GetApi(ORT_API_VERSION));
+  Ort::UnownedSessionOptions session_options(options);
+
+  static CustomOpOpenVINO c_CustomOpOpenVINO(Ort::ConstSessionOptions{options});
```

Comment:
Here, we're storing a const pointer to the session options in the custom op, so that we can query session configs on kernel creation. _Can I get opinions on this please?_ 

**I don't think it's a good idea** to store a pointer to the session options in a custom op due to lifetime constraints. Specifically, a custom op must outlive the ORT session in order to ensure that the custom op is still valid when ORT calls the custom op's [KernelDestroy callback](https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/session/custom_ops.cc#L128).

So, storing a pointer to an object you'll outlive (as is done here) is opening the door to really nasty bugs if, for example, the custom op destructor tried to access the invalid session options for some unlikely reason.

**One alternative** is to not store the session options in the custom op. Instead, we can immediately get the custom op's config options from the session options and pass them down to the custom op:

```cpp
  std::unordered_map<std::string, std::string> op_session_configs;
  session_options.GetSessionConfigs(op_session_configs);

  static CustomOpOpenVINO c_CustomOpOpenVINO(op_session_configs);
```

However, the drawback is that the app must set all session config values _before_ calling the `RegisterCustomOpsLibrary` API. For example, the following works:

```cpp
    Ort::SessionOptions session_opts;
    session_opts.AddCustomOpConfigEntry("OpenVINO_Wrapper", "device_type", "CPU");  // FIRST
    lib_handle = session_opts.RegisterCustomOpsLibrary(lib_name);  // SECOND
    
   Ort::Session session(*ort_env, CUSTOM_OP_OPENVINO_WRAPPER_LIB_TEST_MODEL_URI, session_opts);
```

But the following will not work. The custom op will not see any configs.
```cpp
    Ort::SessionOptions session_opts;
    lib_handle = session_opts.RegisterCustomOpsLibrary(lib_name);  // FIRST (oops)
    session_opts.AddCustomOpConfigEntry("OpenVINO_Wrapper", "device_type", "CPU");  // SECOND
    
   Ort::Session session(*ort_env, CUSTOM_OP_OPENVINO_WRAPPER_LIB_TEST_MODEL_URI, session_opts);
```

If we don't care about imposing this ordering, then this is the better approach, but it seems error-prone to me. Thoughts?

Reply:
For one thing, you do not need two separate things such as UnownedSessionOptions and ConstSessionOptions. The latter exposes only const interfaces, but with regards to ownership, there is not a difference.

Now, with regards to ownership. Typically, the options are read and discard, so the question does not arise. The session itself makes a copy of the options. We can add a `Clone()` method that is already provided for some classes and that can be stored, but not statically, I suggest the actual op class (just thinking on a go).

I am not sure how to solve the problem of setting all of the options before the op is registed, since the code flow only enforces to set all of the options before the actual session is created. One idea is to create a separate class for it.

Reply:
I'm still thinking about how to ensure options are set before op is registered. Here are a few ideas I'm considering:

### CustomOp class
```c++
Ort::CustomOpLibrary op_lib(lib_name);  // Aside: RAII should take care of calling dlclose or FreeLibrary on the lib_handle

op_lib.AddConfig("OpenVINO_Wrapper", "device_type", "CPU");  // Call to set configs
op_lib.AddConfig(...);

op_lib.Register(session_options);  // Wraps C API call to RegisterCustomOpsLibrary()

// op_lib.AddConfig(...);   // This would be invalid and throw an exception??
```

### Pass config options to Register call

This idea requires a new overload of the Ort::SessionOptions::RegisterCustomOpsLibrary method that takes the custom op options as a parameter.

Ex:
```c++
Ort::CustomOpOptions op_options;

op_options.Add("OpenVINO_Wrapper", "device_type", "CPU");
op_options.Add(...);

// This is a new overload.
// Internally, this method sets configs before calling the C API to register the custom op.
void* lib_handle = session_opts.RegisterCustomOpsLibrary(lib_name, op_options);

```

Note that these implementations only help users of the C++ API; C API users still need to set configs before registering the custom op.

I'll update this comment thread with other ideas as I go along.

Reply:
I like the idea of CustomOp class bc it adds RAII and we need that. The name should be changed. `CustomOpLibrary` perhaps?

Make sure the class is movable, so it can be easily stored as a member some place, not just on the stack.

The fact, that these only help to C++ users makes sense, bc this is part of the value C++ API delivers.

Reply:
I added a CustomOpLibrary class. Example use:

```c++
  auto release_lib_func = [](void* lib_handle) {
#ifdef _WIN32
    bool success = ::FreeLibrary(reinterpret_cast<HMODULE>(lib_handle));
    ORT_ENFORCE(success, "Error while closing custom op OpenVINO wrapper shared library");
#else
    int retval = dlclose(lib_handle);
    ORT_ENFORCE(retval == 0, "Error while closing custom op OpenVINO wrapper shared library");
#endif
  };

  Ort::CustomOpLibrary custom_op_lib(lib_name, release_lib_func);
  {
    Ort::SessionOptions session_opts;
    custom_op_lib.Register(session_opts.GetUnowned(), {{"OpenVINO_Wrapper", {{"device_type", "CPU"}}}});
    // ...
  }

```

Reply:
Due to the introduction of a new API for registering custom op libraries that manages the lifetime of lib handles, the above has been simplified to:

```c++
  Ort::SessionOptions session_opts;
  Ort::CustomOpConfigs custom_op_configs;

  custom_op_configs.AddConfig("OpenVINO_Wrapper", "device_type", "CPU");
  session_opts.RegisterCustomOpsLibrary(lib_name, custom_op_configs);
```